mri/reconstruct/utils.py

```diff
@@ -170,10 +171,13 @@ def generate_operators(data, wavelet_name, samples, mu=1e-06, nb_scales=4,
         raise ValueError("At the moment, this functuion only supports 2D "
                          "data.")
     # Define the linear/fourier operators
-    linear_op = WaveletN(
-        nb_scale=nb_scales,
-        wavelet_name=wavelet_name,
-        padding_mode=padding_mode)
+    if wavelet_name != 'UndecimatedBiOrthogonalTransform':
```

Comment:
what is this?

Reply:
This was the only way I could check for undecimated wavelet and provide a different implementation through Modopt. Any update to make this more scalable would be helpful.

Reply:
you probably have the list of names corresponding to `pywt` or isap somewhere in `pysap`. Use that to determine what constructor to chose.

Found it: `pysap.wavelist()` (I just did `dir(pysap)`)

Reply:
No, theres a catch. 
Modopt works by sending wavelet ID. 
So in my opinion the best solution under given constraints that we dont modify pySAP or sparse2D or Modopt would be:
* Create a map between the wavelet names from pysap.wavelist and wavelet_id (see help in `mr_transform`)
* Allow user to choose between the Modopt or pysap implementation. As there is not clear reason why the WaveletUD would not support for decimated wavelets also.


For future and best fix:
* Sparse2D needs to provide a way to instance wavelets from wavelet name just like pysap, or it needs to have a way to map it internally.
* Modopt need to call right or map things right.

Reply:
What you can do is allow the user to specify the name. 
If the name belongs to the list of `pysap` wavelets:
 - if it's `pywt` you implement with `WaveletN`
 - it it's in `isap` you throw a warning and implement with `WaveletN`
If it doesn't it means it belongs to the `mr_transform` ids, so you implement with `WaveletUD2`.
In the example you would just have to specify to which transform you are pointing with an explanation saying it's a temporary fix (and that not all wavelets of `mr_transform` are supported at this point).

Reply:
But then we still need a map to convert names to wavelet IDs , am I right?

Reply:
no, if you want to use the `ModOpt` you have to provide the wavelet id as the wavelet name.
Since this is a temporary fix it's ok. We can remind somewhere that a call to `mr_transform`'s help will give this map.

Reply:
Hmm, dirty fix, but I am ready to go with it for now

Reply:
Ahh I think I missed this last time. I dont think I agree with providing `pywt` or `isap` as input parameter.
A simple approach would be:
try:
    WaveletN(wavelet_name=wavelet_name)
except: 
    WaveletUD2(wavelet_id=wavelet_name)
end

I am trying to get rid of this difference between pywt or isap implementation, as the wavelet name clearly distinguishes it. (pywt usually have them as short forms)

For now, Ill fix based on my above proposal, let me know if you have any other points or drawbacks in my approach.

Reply:
yes I think it's fine for now, since it's just temporary anyway